locator/topology.hh

```diff
@@ -12,37 +12,128 @@
 #include <unordered_map>
+#include <iostream>
 
 #include <seastar/core/future.hh>
 #include <seastar/core/sstring.hh>
 #include <seastar/util/bool_class.hh>
 
 #include "locator/types.hh"
 #include "inet_address_vectors.hh"
-#include "utils/fb_utilities.hh"
 
 using namespace seastar;
 
 namespace locator {
 
+class topology;
+
+class node : public enable_lw_shared_from_this<node> {
+public:
+    using local = bool_class<struct local_tag>;
+
+private:
+    host_id _host_id;
+    inet_address _endpoint;
+    endpoint_dc_rack _dc_rack;
+
+    // Private state of this instance.
+    // May change across topology generations
+    local _is_local;
+
+    friend class topology;
+public:
+    node(host_id id, inet_address endpoint, endpoint_dc_rack dc_rack, local is_local);
+
+    node(const node&) = default;
+    node(node&&) = delete;
+
+    const host_id& host_id() const noexcept {
+        return _host_id;
+    }
+
+    const inet_address& endpoint() const noexcept {
+        return _endpoint;
+    }
+
+    const endpoint_dc_rack& dc_rack() const noexcept {
+        return _dc_rack;
+    }
+
+    local is_local() const noexcept { return _is_local; }
+};
+
+using node_ptr = lw_shared_ptr<const node>;
+using node_set = std::unordered_set<node_ptr>;
+
 class topology {
 public:
     struct config {
+        host_id local_host_id;
+        inet_address local_endpoint;
         endpoint_dc_rack local_dc_rack;
         bool disable_proximity_sorting = false;
     };
     topology(config cfg);
-    topology(topology&&) = default;
+    topology(topology&&) noexcept;
 
     topology& operator=(topology&&) = default;
 
     future<topology> clone_gently() const;
     future<> clear_gently() noexcept;
 
+public:
+    const node_ptr& local_node() const noexcept {
+        return _local_node;
+    }
+
+    // Adds a node with given host_id, endpoint, and DC/rack.
+    node_ptr add_node(host_id id, const inet_address& ep, const endpoint_dc_rack& dr);
+
+    // Optionally updates node's current host_id, endpoint, or DC/rack.
+    // Note: the host_id may be updated from null to non-null after a new node gets a new, random host_id,
+    // or a peer node host_id may be updated when the node is replaced with another node using the same ip address.
+    node_ptr update_node(node_ptr node, std::optional<host_id> opt_id, std::optional<inet_address> opt_ep, std::optional<endpoint_dc_rack> opt_dr);
+
+    using must_exist = bool_class<struct must_exist_tag>;
+
+    // Removes a node using its host_id
+    //
+    // If host_id is not found and must_exist is true:
+    //   the function throws internal error.
+    void remove_node(host_id id, must_exist must_exist = must_exist::no);
+
+    // Finds a node by its host_id
+    //
+    // If host_id is not found:
+    //   the function throws internal error if must_exist is true,
+    //   or returns nullptr otherwise.
+    node_ptr find_node(host_id id, must_exist must_exist = must_exist::no) const noexcept;
```

Comment:
Similar here.

Reply:
The logic is that altering the behavior in such a fundamental way shouldn't be done via a parameter.

Reply:
Makes sense

Reply:
In v8 `must_exist` is gone and `find_node` just returns nullptr if not found and it's up to the callers to handle that.